File: server/src/main/com/broll/gainea/server/GaineaServer.java
package com.broll.gainea.server;

import com.broll.gainea.NetworkSetup;
import com.broll.gainea.server.core.GameContainer;
import com.broll.gainea.server.core.map.Location;
import com.broll.gainea.server.init.LobbyData;
import com.broll.gainea.server.init.PlayerData;
import com.broll.gainea.server.init.ServerSetup;
import com.broll.networklib.server.LobbyGameServer;
import com.broll.networklib.server.LobbyServerCLI;
import com.broll.networklib.server.impl.ServerLobby;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Optional;
import java.util.stream.Collectors;

public class GaineaServer {

    private final static Logger Log = LoggerFactory.getLogger(GaineaServer.class);

    private LobbyGameServer<LobbyData, PlayerData> server;

    public GaineaServer(String version) {
        com.esotericsoftware.minlog.Log.INFO();
        Log.info("Start Gainea Server " + version);
        server = new LobbyGameServer<>("GaineaServer", NetworkSetup::registerNetwork);
        server.setVersion(version);
        ServerSetup.setup(server);
        server.open();
        server.addListener(new ServerStatistic(server));
    }

    public void appendCLI() {
        LobbyServerCLI.open(server, nextTurn(), gameInfo(), giveCard());
    }

    private LobbyServerCLI.CliCommand giveCard() {
        return LobbyServerCLI.cmd("givecard", "gives current player a specific card", options -> {
            if (options.size() < 2) {
                System.err.println("givecard needs to be followed by a lobby id and card picture id");
                return;
            }
            int id = Integer.parseInt(options.get(0));
            int picId = Integer.parseInt(options.get(1));
            getGame(id).ifPresent(game -> {
                game.getCardStorage().getAllCards().stream().filter(it -> it.getPicture() == picId)
                        .findFirst().ifPresent(card ->
                                game.getCurrentPlayer().getCardHandler().receiveCard(card)
                        );
            });
        });
    }

    private LobbyServerCLI.CliCommand nextTurn() {
        return LobbyServerCLI.cmd("nextturn", "Ends the current turn", options -> {
            if (options.isEmpty()) {
                System.err.println("nextturn needs to be followed by a lobby id");
                return;
            }
            int id = Integer.parseInt(options.get(0));
            getGame(id).ifPresent(game -> {
                game.getBattleHandler().reset();
                game.getReactionHandler().getActionHandlers().getReactionActions().endTurn();
                print("Ended turn");
            });
        });
    }

    private LobbyServerCLI.CliCommand gameInfo() {
        return LobbyServerCLI.cmd("game", "Info about a running game", options -> {
            if (options.isEmpty()) {
                System.err.println("game needs to be followed by a lobby id");
                return;
            }
            int id = Integer.parseInt(options.get(0));
            getGame(id).ifPresent(game -> {
                LobbyData settings = game.getGameSettings();
                print("> Game Settings: Map:" + settings.getExpansionSetting().getName() + " Goals:" + settings.getGoalTypes().getName() + " PointLimit:" + settings.getPointLimit() + " MonsterCount:" + settings.getMonsterCount()
                        + " StartGoals:" + settings.getStartGoals() + " StartLocations:" + settings.getStartLocations());
                print("> Round:" + game.getRounds() + " Turn:" + (game.getCurrentTurn() + 1) + "/" + game.getAllPlayers().size() + " Player:" + (game.getCurrentTurn() >= 0 ? game.getCurrentPlayer().getServerPlayer().getName() : "None"));
                print("> State: ProcessingCore.Busy=" + game.getProcessingCore().isBusy() + " BattleHandler.Active=" + game.getBattleHandler().isBattleActive());
                print("> Game Objects [" + game.getObjects().size() + "]:");
                game.getObjects().forEach(object -> print(">> " + object.toString()));
                game.getAllPlayers().forEach(player -> {
                    print("> Player [" + player.toString() + " Points:" + player.getGoalHandler().getScore()
                            + " Stars:" + player.getGoalHandler().getStars() + " Online=" + player.getServerPlayer().isOnline() + "]:");
                    print(">> Goals (" + player.getGoalHandler().getGoals().size() + "):" + player.getGoalHandler().getGoals().stream().map(Object::toString).collect(Collectors.joining(",")));
                    print(">> Cards (" + player.getCardHandler().getCards().size() + "): " + player.getCardHandler().getCards().stream().map(Object::toString).collect(Collectors.joining(",")));
                    print(">> Controlled Locations: [" + player.getControlledLocations().stream().map(Location::toString).collect(Collectors.joining(",")) + "]");
                    print(">> Objects [" + player.getUnits().size() + "]:");
                    player.getUnits().forEach(object -> print(">>> " + object.toString()));
                });
            });
        });
    }

    private void print(String text) {
        System.out.println(text);
    }

    private Optional<GameContainer> getGame(int lobbyId) {
        ServerLobby<LobbyData, PlayerData> lobby = server.getLobbyHandler().getLobby(lobbyId);
        if (lobby == null) {
            System.err.println("Lobby #" + lobbyId + " not found!");
        } else {
            GameContainer game = lobby.getData().getGame();
            if (game == null) {
                System.err.println("Lobby #" + lobbyId + " has no active game!");
            } else {
                return Optional.of(game);
            }
        }
        return Optional.empty();
    }
}

File: server/src/main/com/broll/gainea/server/GaineaServer.kt
package com.broll.gainea.server

import com.broll.gainea.NetworkSetup
import com.broll.gainea.server.core.Game
import com.broll.gainea.server.core.events.RandomEventContainer
import com.broll.gainea.server.core.objects.MapObject
import com.broll.gainea.server.core.player.Player
import com.broll.gainea.server.core.utils.endTurn
import com.broll.gainea.server.init.LobbyData
import com.broll.gainea.server.init.PlayerData
import com.broll.gainea.server.init.ServerSetup
import com.broll.gainea.server.sites.CustomLobbySite
import com.broll.networklib.server.LobbyGameServer
import com.broll.networklib.server.LobbyServerCLI
import com.broll.networklib.server.LobbyServerCLI.CliCommand
import org.slf4j.LoggerFactory
import java.util.function.Consumer

class GaineaServer(version: String) {
    val server: LobbyGameServer<LobbyData, PlayerData>

    init {
        com.esotericsoftware.minlog.Log.INFO()
        Log.info("Start Gainea Server $version")
        server =
            LobbyGameServer("GaineaServer", CustomLobbySite()) { NetworkSetup.registerNetwork(it) }
        server.setVersion(version)
        ServerSetup.setup(server)
        server.open()
        server.addListener(ServerStatistic(server))
    }

    fun appendCLI() {
        LobbyServerCLI.open(server, nextTurn(), triggerEvent(), gameInfo(), giveCard())
    }

    private fun giveCard(): CliCommand {
        return LobbyServerCLI.cmd(
            "givecard",
            "gives current player a specific card"
        ) { options: List<String> ->
            if (options.size < 2) {
                System.err.println("givecard needs to be followed by a lobby id and card picture id")
                return@cmd
            }
            val id = options[0].toInt()
            val picId = options[1].toInt()
            getGame(id)?.let { game ->
                game.cardStorage.allCards.find { card -> card.picture == picId }?.let { card ->
                    game.currentPlayer.cardHandler.receiveCard(card)
                }
            }
        }
    }

    private fun triggerEvent(): CliCommand {
        return LobbyServerCLI.cmd("event", "triggers random event") { options: List<String> ->
            if (options.size < 1) {
                System.err.println("event must be followed by a lobby id")
                return@cmd
            }
            val id = options[0].toInt()
            getGame(id)?.let { game ->
                RandomEventContainer().run(game)
            }
        }
    }

    private fun nextTurn(): CliCommand {
        return LobbyServerCLI.cmd("nextturn", "Ends the current turn") { options: List<String> ->
            if (options.isEmpty()) {
                System.err.println("nextturn needs to be followed by a lobby id")
                return@cmd
            }
            val id = options[0].toInt()
            getGame(id)?.let { game ->
                game.battleHandler.reset()
                game.endTurn()
                print("Ended turn")
            }
        }
    }

    private fun gameInfo(): CliCommand {
        return LobbyServerCLI.cmd("game", "Info about a running game") { options: List<String> ->
            if (options.isEmpty()) {
                System.err.println("game needs to be followed by a lobby id")
                return@cmd
            }
            val id = options[0].toInt()
            getGame(id)?.let { game ->
                val settings = game.gameSettings
                print(
                    "> Game Settings: Map:" + settings.expansionSetting.getName() + " Goals:" + settings.goalTypes.goalName + " PointLimit:" + settings.pointLimit + " MonsterCount:" + settings.monsterCount
                            + " StartGoals:" + settings.startGoals + " StartLocations:" + settings.startLocations
                )
                print("> Round:" + game.rounds + " Turn:" + (game.currentTurn + 1) + "/" + game.allPlayers.size + " Player:" + if (game.currentTurn >= 0) game.currentPlayer.serverPlayer.name else "None")
                print("> State: ProcessingCore.Busy=" + game.processingCore.isBusy + " BattleHandler.Active=" + game.battleHandler.isBattleActive)
                print("> Game Objects [" + game.objects.size + "]:")
                game.objects.forEach(Consumer { `object`: MapObject? -> print(">> " + `object`.toString()) })
                game.allPlayers.forEach(Consumer { player: Player ->
                    print(
                        "> Player [" + player.toString() + " Points:" + player.goalHandler.score
                                + " Stars:" + player.goalHandler.stars + " Online=" + player.serverPlayer.isOnline + "]:"
                    )
                    print(
                        ">> Goals (" + player.goalHandler.goals.size + "):" + player.goalHandler.goals.joinToString(
                            ","
                        )
                    )
                    print(
                        ">> Cards (" + player.cardHandler.cards.size + "): " + player.cardHandler.cards.joinToString(
                            ","
                        )
                    )
                    print(">> Controlled Locations: [" + player.controlledLocations.joinToString(",") + "]")
                    print(">> Objects [" + player.units.size + "]:")
                    player.units.forEach { print(">>> $it") }
                })
            }
        }
    }

    private fun print(text: String) {
        println(text)
    }

    private fun getGame(lobbyId: Int): Game? {
        val lobby = server.lobbyHandler.getLobby(lobbyId)
        if (lobby == null) {
            System.err.println("Lobby #$lobbyId not found!")
        } else {
            val game = lobby.data.game
            if (game == null) {
                System.err.println("Lobby #$lobbyId has no active game!")
            } else {
                return game
            }
        }
        return null
    }

    companion object {
        private val Log = LoggerFactory.getLogger(GaineaServer::class.java)
    }
}
